apps/dhparam.c

```diff
@@ -240,20 +240,20 @@ int dhparam_main(int argc, char **argv)
                        num, g);
             BIO_printf(bio_err, "This is going to take a long time\n");
             if (!EVP_PKEY_paramgen_init(ctx)) {
+                ERR_print_errors(bio_err);
                 BIO_printf(bio_err,
                            "Error, unable to initialise DH param generation\n");
-                ERR_print_errors(bio_err);
```

Comment:
Why the order change?

Reply:
> Why the order change?

Because usually the low-level errors are printed before/above the high-level ones.
Not following this pattern (which is natural due to causality) is confusing.

In the cases handled here (`dhparam.c` and `dsaparam.c`) there is a further motivation, 
which I mentioned explicitly in the commit message: consistency.
Both apps were partly calling `ERR_print_errors()` before `BIO_printf()` and partly after.

Reply:
If you look around in other commands, I think you will discover that they print the application level message first and then dump the error queue.  As a matter of fact, some of them have this generic thing at the end of their main function:

``` C
    if (ret != 0)
        ERR_print_errors(bio_err);
```

... which I actually find more appealing than littering the source with a gazillion  `ERR_print_errors(bio_err)` calls.

So in terms of consistency, what you're doing is not.

Reply:
I see your point with the consistency with other apps.

>  As a matter of fact, some of them have this generic thing at the end of their main function:

    if (ret != 0)
        ERR_print_errors(bio_err);

> ... which I actually find more appealing than littering the source with a gazillion ERR_print_errors(bio_err) calls.

This is really a strong argument. Done for these two apps as well.
Also added a couple of missing high-level error messages.